File: leetcode/0161-one-edit-distance/Solution.java
class Solution {
    public boolean isOneEditDistance(String s, String t) {
        if (Math.abs(s.length() - t.length()) > 1) {
            return false;
        }

        if (s.length() == t.length()) {
            int differences = 0;
            for (int k = 0; k < s.length(); k++) {
                if (s.charAt(k) != t.charAt(k)) {
                    differences++;
                }
            }
            return differences == 1;
        }

        String shorter = s.length() < t.length() ? s : t;
        String longer = s.length() < t.length() ? t : s;

        int i = 0, j = 0;
        boolean foundDifference = false;
        while (i < shorter.length() && j < longer.length()) {
            if (shorter.charAt(i) == longer.charAt(j)) {
                i++;
                j++;
            } else {
                if (foundDifference) {
                    return false;
                }
                foundDifference = true;
                j++;
            }
        }

        return true;
    }
}

File: leetcode/0161-one-edit-distance/solution.cpp
#include <string>
#include <cmath>
using namespace std;

class Solution {
public:
    bool isOneEditDistance(string s, string t) {
        if (abs((int)s.size() - (int)t.size()) > 1) {
            return false;
        }

        if (s.size() == t.size()) {
            int differences = 0;
            for (size_t k = 0; k < s.size(); k++) {
                if (s[k] != t[k]) differences++;
            }
            return differences == 1;
        }

        string shorter = s.size() < t.size() ? s : t;
        string longer = s.size() < t.size() ? t : s;

        size_t i = 0, j = 0;
        bool foundDifference = false;
        while (i < shorter.size() && j < longer.size()) {
            if (shorter[i] == longer[j]) {
                i++;
                j++;
            } else {
                if (foundDifference) return false;
                foundDifference = true;
                j++;
            }
        }

        return true;
    }
};
